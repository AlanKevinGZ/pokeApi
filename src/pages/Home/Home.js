import React, { useEffect } from 'react'
import Header from '../../components/Header/Header'
// ========================================
// IMPORTAR HOOKS DE REACT-REDUX
// ========================================
import { useDispatch, useSelector } from 'react-redux'

// useDispatch: para ENVIAR acciones (como fetchPokemons)
// useSelector: para LEER datos del store
import { fetchPokemons, selectPokemon } from '../../store/PokemonSlice'
import { Container, PokemonButton, PokemonItem, PokemonList, Section, SelectedPokemon, Title } from './style/HomeStyle'
import { useNavigate } from 'react-router-dom'

// ========================================
// IMPORTAR ACCIONES DEL SLICE
// ========================================

// fetchPokemons: acción asíncrona para traer pokémons de la API
// selectPokemon: acción para seleccionar un pokémon

function Home() {
  // ========================================
  // 1. CONFIGURAR DISPATCH
  // ========================================
  // dispatch es como un "mensajero" que envía acciones al store
  const dispatch = useDispatch()
 const navigate = useNavigate(); 
  // ========================================
  // 2. LEER DATOS DEL STORE CON useSelector
  // ========================================
  // Extraemos los datos que necesitamos del store
  // state.pokemon viene de la clave que pusiste en configureStore
  const { list, selected, loading, error } = useSelector((state) => state.pokemon)
  // list: array de pokémons
  // selected: pokémon seleccionado (o null)
  // loading: true/false si está cargando
  // error: mensaje de error (o null)

  // ========================================
  // 3. LLAMAR LA API CUANDO EL COMPONENTE SE MONTA
  // ========================================
  useEffect(() => {
    // dispatch(fetchPokemons()) ejecuta la acción asíncrona
    // Esto automáticamente:
    // 1. Pone loading en true
    // 2. Llama a la API
    // 3. Guarda los pokémons en list
    dispatch(fetchPokemons())

    
  }, [dispatch]) // Se ejecuta solo una vez al montar el componente

  // ========================================
  // 4. FUNCIÓN PARA MANEJAR CLICK EN POKÉMON
  // ========================================
  const handleSelectPokemon = (pokemon) => {
    // Enviamos la acción selectPokemon con el pokémon como payload
    dispatch(selectPokemon(pokemon))
  }

  const moreInfo = (name) => {
    navigate(`/more/${name}`)
    
  }

  // ========================================
  // 5. RENDERIZADO CONDICIONAL
  // ========================================
  
  // Si está cargando, mostramos un mensaje
  if (loading) {
    return (
      <div>
        <Header title="poke Api"></Header>
        <p>Cargando pokémons...</p>
      </div>
    )
  }

  // Si hay un error, lo mostramos
  if (error) {
    return (
      <div>
        <Header title="poke Api"></Header>
        <p style={{ color: 'red' }}>Error: {error}</p>
      </div>
    )
  }

  // ========================================
  // 6. RENDERIZADO NORMAL
  // ========================================
  return (
    <Container>
      <Header title="Poké API" />

      {/* Pokémon seleccionado */}
      {selected && (
        <SelectedPokemon>
          <Title>Pokémon Seleccionado:</Title>
          <p><strong>{selected.name}</strong></p>
          <PokemonButton selected onClick={()=>moreInfo(selected.name)}>Más información</PokemonButton>
        </SelectedPokemon>
      )}

      {/* Lista de pokémons */}
      <Section>
        <Title>Lista de Pokémons:</Title>
        <PokemonList>
          {list.map((pokemon) => (
            <PokemonItem key={pokemon.name}>
              <PokemonButton
                selected={selected?.name === pokemon.name}
                onClick={() => handleSelectPokemon(pokemon)}
              >
                {pokemon.name}
              </PokemonButton>
            </PokemonItem>
          ))}
        </PokemonList>
      </Section>
    </Container>
  )
}

export default Home