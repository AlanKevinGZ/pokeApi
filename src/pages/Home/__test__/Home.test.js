// src/pages/Home/__test__/Home.test.js
import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import "@testing-library/jest-dom";
import Home from "../Home";
import pokemonReducer from "../../../store/PokemonSlice";


const mockTheme = {
  colors: {
    buttons: '#3498db',
    primary: '#2c3e50',
    secondary: '#ecf0f1',
    text: '#ffffff',
    background: '#f5f5f5',
  },
};

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

jest.mock("../../../components/Header/Header", () => {
  return function MockHeader({ title }) {
    return <div data-testid="header">{title}</div>;
  };
});
global.fetch = jest.fn();

const renderWithProviders = (initialState = {}) => {
  const defaultState = {
    list: [],
    selected: null,
    loading: false,
    error: null,
    pokemonDetails: null,
    loadingDetails: false,
    errorDetails: null,
    ...initialState
  };

  const store = configureStore({
    reducer: {
      pokemon: pokemonReducer,
    },
    preloadedState: {
      pokemon: defaultState
    }
  });

  return {
    ...render(
      <Provider store={store}>
        <ThemeProvider theme={mockTheme}>
          <MemoryRouter>
            <Home />
          </MemoryRouter>
        </ThemeProvider>
      </Provider>
    ),
    store
  };
};


describe("Home Component", () => {

  beforeEach(() => {
    jest.clearAllMocks();
    mockNavigate.mockClear();
   
    global.fetch.mockResolvedValue({
      json: async () => ({ results: [] }),
    });
  });


  test("renders Header component", async () => {
    renderWithProviders();
    
   
    await waitFor(() => {
      expect(screen.getByTestId("header")).toBeInTheDocument();
    });
  });


  test("shows loading message when loading is true", () => {
    renderWithProviders({ loading: true });
    expect(screen.getByText(/cargando pokémons/i)).toBeInTheDocument();
  });


  test("shows error message when error exists", async () => {
   
    global.fetch.mockRejectedValueOnce(new Error("Network error"));
    
    renderWithProviders();
    
  
    await waitFor(() => {
      expect(screen.getByText(/error/i)).toBeInTheDocument();
    });
  });

 
  test("renders list of pokemons", async () => {
   
    global.fetch.mockResolvedValueOnce({
      json: async () => ({
        results: [
          { name: "bulbasaur", url: "https://pokeapi.co/api/v2/pokemon/1/" },
          { name: "charmander", url: "https://pokeapi.co/api/v2/pokemon/4/" },
          { name: "squirtle", url: "https://pokeapi.co/api/v2/pokemon/7/" },
        ]
      }),
    });

    renderWithProviders();

   
    await waitFor(() => {
      expect(screen.getByText(/bulbasaur/i)).toBeInTheDocument();
    });
    
    expect(screen.getByText(/charmander/i)).toBeInTheDocument();
    expect(screen.getByText(/squirtle/i)).toBeInTheDocument();
  });


  test("selects a pokemon when clicked", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => ({
        results: [
          { name: "pikachu", url: "https://pokeapi.co/api/v2/pokemon/25/" }
        ]
      }),
    });

    renderWithProviders();

    
    await waitFor(() => {
      expect(screen.getByText(/pikachu/i)).toBeInTheDocument();
    });

    const pikachuButton = screen.getByRole('button', { name: /pikachu/i });
    fireEvent.click(pikachuButton);
    
    
    await waitFor(() => {
      expect(screen.getByText(/pokémon seleccionado/i)).toBeInTheDocument();
    });
  });


  test("shows selected pokemon section when selected", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => ({
        results: [
          { name: "charmander", url: "https://pokeapi.co/api/v2/pokemon/4/" }
        ]
      }),
    });

    renderWithProviders({
      selected: { name: "charmander", url: "https://pokeapi.co/api/v2/pokemon/4/" },
    });

    await waitFor(() => {
      expect(screen.getByText(/pokémon seleccionado/i)).toBeInTheDocument();
    });
  });

  test("shows more info button when pokemon selected", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => ({
        results: [
          { name: "squirtle", url: "https://pokeapi.co/api/v2/pokemon/7/" }
        ]
      }),
    });

    renderWithProviders({
      selected: { name: "squirtle", url: "https://pokeapi.co/api/v2/pokemon/7/" },
    });

    await waitFor(() => {
      expect(screen.getByText(/más información/i)).toBeInTheDocument();
    });
  });


  test("renders correctly with empty list", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => ({ results: [] }),
    });

    renderWithProviders();
    
    await waitFor(() => {
      expect(screen.getByTestId("header")).toBeInTheDocument();
    });
    
    expect(screen.getByText(/lista de pokémons/i)).toBeInTheDocument();
  });


  test("matches snapshot", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => ({
        results: [
          { name: "pikachu", url: "https://pokeapi.co/api/v2/pokemon/25/" },
        ]
      }),
    });

    const { container } = renderWithProviders();

    await waitFor(() => {
      expect(screen.getByText(/pikachu/i)).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });

  test("navigates to more info page when clicking more info button", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => ({
        results: [
          { name: "pikachu", url: "https://pokeapi.co/api/v2/pokemon/25/" }
        ]
      }),
    });

    renderWithProviders({
      selected: { name: "pikachu", url: "https://pokeapi.co/api/v2/pokemon/25/" },
    });

   
    await waitFor(() => {
      expect(screen.getByText(/más información/i)).toBeInTheDocument();
    });

 
    const moreInfoButton = screen.getByText(/más información/i);
    fireEvent.click(moreInfoButton);


    expect(mockNavigate).toHaveBeenCalledWith('/more/pikachu');
    expect(mockNavigate).toHaveBeenCalledTimes(1);
  });
});