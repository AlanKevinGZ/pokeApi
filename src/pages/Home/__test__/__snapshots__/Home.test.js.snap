// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component matches snapshot 1`] = `
<div>
  <div
    class="sc-bRKDuR jnvOHb"
  >
    <div
      data-testid="header"
    >
      Poké API
    </div>
    <div
      class="sc-hvigdm caMVCR"
    >
      <h2
        class="sc-fhHczv kzeObA"
      >
        Lista de Pokémons:
      </h2>
      <ul
        class="sc-ggWZvA fhCMza"
      >
        <li
          class="sc-dTvVRJ dtDCyz"
        >
          <button
            class="sc-jwTyAe jRayZA"
          >
            pikachu
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
